### PR TITLE
[ADD] stock_picking_delivery_info_computation: Improve on calculation…

### PR DESCRIPTION
**Pull request by victor-ditay**
… weight and volume

### CODE REVIEW

File: stock_picking_delivery_info_computation/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Tecnativa - Victor M.M. Torres
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+{
+    'name': 'Stock Picking Delivery Info Computation',
+    'summary': '',
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/__manifest__.py]: Remove this key if empty

File: stock_picking_delivery_info_computation/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    'version': '10.0.1.0.0',
+    'category': 'Sales, Stock, Delivery',
+    'website': 'https://github.com/OCA/delivery-carrier',
+    'author': 'Tecnativa, Odoo Community Association (OCA)',
+    'license': 'LGPL-3',
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/__manifest__.py]: Why `LGPL-3`? Put `AGPL-3`

File: stock_picking_delivery_info_computation/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    'installable': True,
+    'depends': [
+        'delivery',
+    ],
+    'data': [
+        'views/delivery_view.xml',
+    ],
+    'demo': [
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/__manifest__.py]: Don't put empty sections

File: stock_picking_delivery_info_computation/models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Tecnativa - Victor M.M. Torres
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/__init__.py]: Don't put copyright in init files

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+# Copyright 2019 Tecnativa - Victor M.M. Torres
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    volume = fields.Float(
+        compute='_calculate_volume',
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: This field mustn't be computed.

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                else:
+                    for pack in rec.pack_operation_product_ids:
+                        weight += (
+                            pack.product_qty * pack.product_id.weight)
+            elif rec.move_lines:
+                for move in rec.move_lines:
+                    weight += (
+                        move.product_uom_qty * move.product_id.weight)
+            rec.weight = weight
+        if not weight:
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: This is not correct. The super condition is that exists pack operations

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                if any(
+                    pack.qty_done > 0 for pack in rec.mapped(
+                        "pack_operation_product_ids")):
+                    for pack in rec.pack_operation_product_ids:
+                        weight += pack.qty_done * pack.product_id.weight
+                else:
+                    for pack in rec.pack_operation_product_ids:
+                        weight += (
+                            pack.product_qty * pack.product_id.weight)
+            elif rec.move_lines:
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: This never should look on `move_lines`. super does that

**Comment by victor-ditay** [comment]
cc @Tecnativa @pedrobaeza

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                        "pack_operation_product_ids")):
+                    for pack in rec.pack_operation_product_ids:
+                        weight += pack.qty_done * pack.product_id.weight
+                else:
+                    for pack in rec.pack_operation_product_ids:
+                        weight += (
+                            pack.product_qty * pack.product_id.weight)
+                rec.weight = weight
+            else:
+                super(StockPicking, self)._cal_weight()
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]:
```suggestion
                super(StockPicking, rec)._cal_weight()
```

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    @api.multi
+    def action_calculate_volume(self):
+        total_volume = 0.0
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: This variable must be reset on each step of the loop, not globally

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    @api.depends(
+        'product_id',
+        'move_lines',
+        'pack_operation_product_ids')
+    def _cal_weight(self):
+        weight = 0
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: This variable must be reset on each step of the loop, not globally

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+# Copyright 2019 Tecnativa - Victor M.M. Torres
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, models
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    @api.depends(
+        'product_id',
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: This field doesn't exist at picking level

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        'move_lines',
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: This depends is already defined in super, so no need to repeat.

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        'pack_operation_product_ids')
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: You must add the subfields that influences in this compute:
```suggestion
        'pack_operation_product_ids',
        'pack_operation_product_ids.product_id',
        'pack_operation_product_ids.product_qty',
        'pack_operation_product_ids.qty_done',
    )
```

File: stock_picking_delivery_info_computation/models/stock_picking.py (hunk 8 above)
Comment (pedrobaeza) [reply stock_picking_delivery_info_computation/models/stock_picking.py]: But better to do it on batch:
```python
with_pack_ops = self.filtered('pack_operation_product_ids')
for rec in with_pack_ops:
    ...
super(StockPicking, self - with_pack_ops)._cal_weight()
```

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    def _cal_weight(self):
+        weight = 0
+        for rec in self:
+            if rec.pack_operation_product_ids:
+                if any(
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: Better code:
```python
has_done = any(rec.mapped('pack_operation_product_ids.qty_done'))
field = 'qty_done' if has_done else 'product_qty'
rec.weight = sum(rec.pack_operation_product_ids.mapped(
    lambda x: x[field] * x.product_id.weight
))
```
`any` already checks the values non zeroed, so it's enough.

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+                rec.weight = weight
+            else:
+                super(StockPicking, self)._cal_weight()
+
+    @api.multi
+    def action_calculate_volume(self):
+        total_volume = 0.0
+        for rec in self:
+            if rec.pack_operation_product_ids:
+                if any(
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: Apply the same logic as explained above

File: stock_picking_delivery_info_computation/models/stock_picking.py [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+                    pack.qty_done > 0 for pack in rec.mapped(
+                        "pack_operation_product_ids")):
+                    for pack in rec.pack_operation_product_ids:
+                        total_volume += pack.qty_done * pack.product_id.volume
+                else:
+                    for pack in rec.pack_operation_product_ids:
+                        total_volume += (
+                            pack.product_qty * pack.product_id.volume)
+            elif rec.move_lines:
+                for move in rec.move_lines:
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/models/stock_picking.py]: Use `mapped`

File: stock_picking_delivery_info_computation/readme/CREDITS.rst [hunk 17]
```diff
@@ -0,0 +1,3 @@
+The development of this module has been financially supported by:
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/readme/CREDITS.rst]: This is not true: we are developers, not funders

File: stock_picking_delivery_info_computation/tests/test_stock_picking_info_computation.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+        self.product_uom_unit = self.env.ref('product.product_uom_unit')
+        self.product_category_5 = self.env.ref('product.product_category_5')
+        self.product_pricelist_0 = self.env.ref('product.list0')
+        self.product_a = self.env['product.product'].create({
+            'name': 'Test product A',
+            'categ_id': self.product_category_5.id,
+            'standard_price': 100.0,
+            'list_price': 200.0,
+            'type': 'consu',
+            'uom_id': self.product_uom_unit.id,
```
Comment (pedrobaeza) [inline stock_picking_delivery_info_computation/tests/test_stock_picking_info_computation.py]: No need of specifying all of these things. Only relevant fields like name, type, weight or volume

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Add usage section explaining about the button for computing the volume, and specify the moments where the volume is automatically computed.

File: stock_picking_delivery_info_computation/models/stock_picking.py (hunk 11 above)
Comment (victor-ditay) [reply stock_picking_delivery_info_computation/models/stock_picking.py]: https://github.com/odoo/odoo/blob/10.0/addons/stock/models/stock_picking.py#L289

Perhaps is not relevant on this computation but the field exist and I keep on the definition as it comes from the standard code: https://github.com/odoo/odoo/blob/10.0/addons/delivery/models/stock_picking.py#L105

But as your following comment https://github.com/OCA/delivery-carrier/pull/208#discussion_r307431468 I don't need to repeat this definition, so I will remove this part

Comment (pedrobaeza) [reply stock_picking_delivery_info_computation/models/stock_picking.py]: Uhm, a related to a one2many field... Not sure how that behaves, but OK, we don't need to repeat it.

**Review by pedrobaeza — APPROVED** [review]
Added missing features + better code:

* Better readme
* Improved tests
* Compute volume on sales order confirmation
* Compute volume on backorder creation
* Better code in some places

**Review by Tardo — APPROVED** [review]
LGTM :+1:, code review

**Comment by rafaelbn** [comment]
@pedrobaeza your merge command was aborted due to failed check(s), which you can inspect on this commit of 10.0-ocabot-merge-pr-208-by-pedrobaeza-bump-no.

**Comment by rafaelbn** [comment]
@pedrobaeza here says . We should delete branch

Closed with unmerged commits 

This pull request is closed, but the Tecnativa:10.0-ADD-stock-picking-delivery-info-computation branch has unmerged commits. You can delete this branch if you wish.
If you wish, you can also delete your fork of OCA/delivery-carrier.

**Comment by pedrobaeza** [comment]
@rafaelbn please read:

Selección_001

**Pull request closed without merging**